LSP workspaces

pascalkuthe opened the pull request:
Supersedes #4439 and (mostly) #3207
Fixes: #3993
Fixes: #4436
fixes #5852
Closes #5919

The aim of this PR is primarily to address the problem from #4439 as described in https://github.com/helix-editor/helix/pull/4439#issuecomment-1345568363. As the solution described there uses a workspace config I ended up implementing support for that as well. The syntax I proposed in https://github.com/helix-editor/helix/pull/4439#issuecomment-1345568363 didn't quite work out as adding a new top-level header (that can be accessed in `helix-view`) like `[workspace]` would have required more changes that didn't seem worth the effort here. Furthermore, the language specialization I proposed in my comment is incompatible with the toml standard (sadly). I ended up with the following solution:

To force LSP to stop searching for roots in the `fuzz` folder you can
add the following `.helix/config.toml` file:
``` toml
[editor]
workspace-lsp-roots = ["fuzz"]
```

To only apply this setting only to `rust` (but not other languages where this might cause problems) you can add the following `.helix/languages.toml` file instead:

``` toml
[[language]]
name = "rust"
workspace-lsp-roots = ["fuzz"]
```

To handle local config I originally wanted to pull in #3207, but when I started reviewing the changes there I quickly found many things that required changes so that I decided to reimplement the functionality here instead. Compared to #3207 this PR does not implement the security setting for workspace config. It's disabled by default anyway so that feature can be added as a separate followup PR (or #3207 could be rebased on this)

Furthermore, this PR correctly handles:
* Respect local config during `:config-reload` and `:set` (would require larger refactor with #3207)
* Full keymap merging of arbitrary nested keybindings (just like merging with the default keymap works on master)
* Unify config loading in `helix-term/src/main.rs` and during config-reload/set

Finally, this PR contains on breaking change: How `.helix` is treated. Currently, we simply merge all `.helix/langauges.toml` from the CWD to the workspace root. This PR changes the workspace lookup so `.helix` acts just like `.git`. It's not intended for marking LSP roots (that's what `workspace-roots` is for) but rather for cases where:
* You are working without `git`
* Monorepos where you want to really only work on part of the repo as a workspace (I am talking HUGE monorepos like the windows monorepo) for all operations (including the file picker)

As the workspace root search always stops at the first `.helix` directory, there is only a single `.helix/config.toml` and `.helix/languages.toml`, so the merging behavior is removed (breaking change I was talking about). In addition to the use-cases listed above this also ensure that it's always clear what the `workspace-roots` setting is referring to: A path relative to the parent of `.helix`. When multiple configs are merged it can be quite confusing what these directories refer to otherwise. Finally, this behavior is also more inline with other editors (like `.vscode` for VSCode).

pascalkuthe commented:
One thing I would like feedback in: Right now the `:change-dirextory` command does not automatically load the local configuration from that folder. 

The reason is that the global and the local configuration are not saved seperatly and therefore can only be reloaded together (so essentially calling `:config-reload`). It would have been yet another (altough subtle) breaking change to automatically call `:reload-config` in changing directories so I held off. To me this feels like a better solution so different `workspace-roots` are picked up automatically. Ofcourse calling :cobfig-reload` manually works but it's easy to forget and will likely cause tickets to be created

pascalkuthe commented:
rebased on master to resolve conflicts, the changes from #5420 actually make this PR a bit smaller/more elegant as the config can be accessed trough the document now.

LeoniePhiline commented:
5083fd11e3bb1a77880aa29bdc1eb2261a961890 works perfectly for my situation described at #5852

LeoniePhiline commented on helix-lsp/src/lib.rs:
```diff
@@ -589,6 +597,53 @@ fn start_client(
+/// * start at `file` (either an absolute path or relative to CWD)
+/// * find the top most directory containing a root_marker
+/// * inside the current workspace
+///     * stop the search at the first root_dir that contains `root` or the workspace (obtained from `helix_core::find_workspace`)
```
```suggestion
///     * stop the search at the first root_dir that contains `file` or the workspace (obtained from `helix_core::find_workspace`)
```

pascalkuthe commented:
I expanded this PR to also fix #5852.

The fundamental issue there was that the workspace root and the lsp root lookup were unified when these are really two separate functions. The LSP lookup always starts at the document while the workspace lookup should always start at the CWD.  The LSP lookup should not look for the workspace root (`.git` or `.helix`) itself because then netsted `.git` repos don't work as the LSP lookup stops too early (which was the cause of #5852). 

With the changes here the workspace lookup and the LSP root lookup are now separate. The workspace lookup just scans up the directory tree from `CWD` till `.git` or `.helix` is found and fallsback to CWD if neither exists.

The LSP root lookup searches for the topmost root_marker from the document.
If the document is inside the workspace folder the search stops if one of the configured `workspace-roots` are discovered or if the workspace root is reached.
Outside of the workspace the search stops at the FS root (still fallsback to the workspace at that case as that seems like the better fallback to me).

LeoniePhiline commented on book/src/languages.md:
```diff
@@ -62,6 +62,7 @@ These configuration keys are available:
 | `grammar`             | The tree-sitter grammar to use (defaults to the value of `name`) |
 | `formatter`           | The formatter for the language, it will take precedence over the lsp when defined. The formatter must be able to take the original file as input from stdin and write the formatted file to stdout |
 | `max-line-length`     | Maximum line length. Used for the `:reflow` command and soft-wrapping           |
+| `workspace-roots`     | Directories relative to the workspace root that are treated as LSP roots. Should only be set in `.helix/config.toml`. Overwrites the setting of the same name in `config.toml` if set. | `` |
```
`workspace-roots` (`Directories...`) means something else than `the workspace root` mentioned in the `Description` column.

Could this better be named `language-server-roots` / `lsp-roots`?

pascalkuthe replied:
Hmm not sure. The reason I called this workspace-roots is that this is supposed to be something that is configured per workspace (in `.helix/config.toml`) so this is meant as `roots` inside the workspace. Maybe `workspace-lsp-roots` works better 🤔

LeoniePhiline replied:
Sounds perfect to me. `workspace-lsp-roots` transmits both the workspace-specificity as well as the lsp-target.

LeoniePhiline commented on helix-core/src/syntax.rs:
```diff
@@ -126,6 +126,9 @@ pub struct LanguageConfiguration {
 
     pub rulers: Option<Vec<u16>>, // if set, override editor's rulers
+
+    /// Hardcoded LSP root direcotries relative to the workspace root like `examples` or `tools/fuzz`
```
```suggestion
    /// Hardcoded LSP root directories relative to the workspace root, like `examples` or `tools/fuzz`. 
    /// Falling back to the current working directory if none are configured.
```

pascalkuthe commented:
While testing this further I noticed that the new solution wasn't working well because we just used the first root we encountered as the LSP workspace/root.

The right solution here is to use one LSP workspace for each separate resolved root. This not just makes my new solution here viable but more importantly also fixes existing issues. Consider the following directory structure with two seperate Cargo workspaces in the same git repo:

```
| - .git
| - workspace1
|   | - Cargo.toml
|   | - src/bar.rs
| - workspace2
|   | - Cargo.toml
|   | - src/foo.rs
```

If you open `workspace1/src/bar.rs` with helix master, RA will be initalized with a single workspace: `workspace1`.
If you open `workspace2/src/foo.rs` afterwards RA will show the file as not being part of any cargo project. [^1]
If you open those two files in reverse order the same problem will occur in reverse (so `bar.rs` would have no LSP).

With this PR the workspaces are both added as an `LSP` workspace and therefore everything works as intended.
However while testing this PR I noticed that RA was indexing a new workspace whenever I opened up a `crates.io` dependency with `gd` and even showed errors in their testsuite (which RA usually ignores for dependencies). The reason is that with these new changes helix now correctly detected those crates `Cargo.toml` and added them as a new workspace.

This is not desirable as dependencies are obiously not supposed to be part of the workspace.
Therefore I have changed the LSP root detection to only create LSP workspaces for files in the current workspace.
All files outside of the current workspace are:

* If there is not LSP yet: Spawn an LSP with `workspace_dirs` set to an empty list, `root_uri` set to `None` and `root_dir` set to CWD (see https://github.com/helix-editor/helix/issues/4436#issuecomment-1288802003, `root_dir` must be set for backwards compatibility)
* If there is already an LSP and it supports multiple workspaces: Files are just added to the LSP without creating a new workspace
* If there is already an LSP and it doesn't support multiple workspaces: Use the LSP with `root_uri` set to `None` (or create it if necessary)

While working on this I ran across #4436. Fixing that issue was easy by creating a special case if the current workspace resolved to the CWD because there is no `.helix` or `.git`:
If no root marker matches for the given file, treat it astough like it lives outside of the current workspace: (described above).

 ^1: Note that this is only true if these two workspaces don't depend on each other. However that is RA specific behavior and not relevant here. RA is able to handle a file belonging to multiple workspaces just fine

LeoniePhiline commented:
fcc445a crashes on me with:

```
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', /home/leonie/Projekte/rust/helix-install/helix-term/src/application.rs:1013:42
stack backtrace:
   0: rust_begin_unwind
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panicking.rs:143:14
   2: core::panicking::panic
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panicking.rs:48:5
   3: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::fold
   4: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter
   5: helix_term::application::Application::handle_editor_event::{{closure}}
   6: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
   7: tokio::runtime::park::CachedParkThread::block_on
   8: tokio::runtime::scheduler::multi_thread::MultiThread::block_on
   9: hx::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

Steps:
1. Open file
2. Press `j` to move cursor down
3. Panic.

jknoetzke commented:
Hi. Silly question. These commits are in the main branch yes ? To test, I simply have to pull ?

LeoniePhiline commented:
@jknoetzke 
```sh
git fetch origin pull/5748/head
git checkout FETCH_HEAD
cargo install --locked --path helix-term
```

pascalkuthe commented:
> [fcc445a](https://github.com/helix-editor/helix/commit/fcc445ac03cb351ae321813a0091d96669c43a9c) crashes on me with:
> 
> ```
> thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', /home/leonie/Projekte/rust/helix-install/helix-term/src/application.rs:1013:42
> stack backtrace:
>    0: rust_begin_unwind
>              at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:584:5
>    1: core::panicking::panic_fmt
>              at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panicking.rs:143:14
>    2: core::panicking::panic
>              at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panicking.rs:48:5
>    3: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::fold
>    4: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter
>    5: helix_term::application::Application::handle_editor_event::{{closure}}
>    6: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
>    7: tokio::runtime::park::CachedParkThread::block_on
>    8: tokio::runtime::scheduler::multi_thread::MultiThread::block_on
>    9: hx::main
> note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
> ```
> 
> Steps:
> 
>     1. Open file
> 
>     2. Press `j` to move cursor down
> 
>     3. Panic.

that's odd it works well with RA or tablo and shouldn't really be happening. Which LS are you using?

LeoniePhiline commented:
Intelephense. I tried to verify by reproducing https://github.com/helix-editor/helix/issues/5852

pascalkuthe commented:
sorry I introduced a stupid bug while clearing up some code and was accidentally still testing with an older version. I have pushed a fix

LeoniePhiline commented:
Crashes when `gd` on a framework class.

```
❯ RUST_BACKTRACE=full hx typo3/public/typo3conf/ext/digibord_authentication/Classes/System/Tca/Element/AbstractElement.php                                                                                                                                                                                      2.39  
thread 'main' panicked at 'Cannot block the current thread from within a runtime. This happens because a function attempted to block the current thread while the thread is being used to drive asynchronous tasks.', helix-lsp/src/client.rs:80:60
stack backtrace:
   0:     0x563e6d30db74 - std::backtrace_rs::backtrace::libunwind::trace::h22893a5306c091b4
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x563e6d30db74 - std::backtrace_rs::backtrace::trace_unsynchronized::h29c3bc6f9e91819d
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x563e6d30db74 - std::sys_common::backtrace::_print_fmt::he497d8a0ec903793
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/sys_common/backtrace.rs:66:5
   3:     0x563e6d30db74 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h9c2a9d2774d81873
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/sys_common/backtrace.rs:45:22
   4:     0x563e6ca3f79c - core::fmt::write::hba4337c43d992f49
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/fmt/mod.rs:1194:17
   5:     0x563e6d306ef5 - std::io::Write::write_fmt::heb73de6e02cfabed
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/io/mod.rs:1655:15
   6:     0x563e6d30fa6e - std::sys_common::backtrace::_print::h63c8b24acdd8e8ce
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/sys_common/backtrace.rs:48:5
   7:     0x563e6d30fa6e - std::sys_common::backtrace::print::h426700d6240cdcc2
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/sys_common/backtrace.rs:35:9
   8:     0x563e6d30fa6e - std::panicking::default_hook::{{closure}}::hc9a76eed0b18f82b
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:295:22
   9:     0x563e6d30f79d - std::panicking::default_hook::h2e88d02087fae196
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:314:9
  10:     0x563e6d3100fb - std::panicking::rust_panic_with_hook::habfdcc2e90f9fd4c
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:702:17
  11:     0x563e6d30ff24 - std::panicking::begin_panic_handler::{{closure}}::he054b2a83a51d2cd
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:588:13
  12:     0x563e6d30e0a4 - std::sys_common::backtrace::__rust_end_short_backtrace::ha48b94ab49b30915
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/sys_common/backtrace.rs:138:18
  13:     0x563e6d30fc8d - rust_begin_unwind
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:584:5
  14:     0x563e6c9ace53 - core::panicking::panic_fmt::h366d3a309ae17c94
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panicking.rs:143:14
  15:     0x563e6ca3d9a1 - core::panicking::panic_display::hd3e89d99fb893792
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panicking.rs:72:5
  16:     0x563e6ca3d94b - core::panicking::panic_str::h98ed104b0c03a108
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panicking.rs:56:5
  17:     0x563e6c9accc6 - core::option::expect_failed::h58d487f78ae9d247
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/option.rs:1874:5
  18:     0x563e6ccff0dc - tokio::future::block_on::block_on::h28e2e7528581cf6d
  19:     0x563e6ccecaf0 - helix_lsp::client::Client::try_add_doc::h56cf9594d9449679
  20:     0x563e6ccc3ae9 - helix_lsp::Registry::get::hb897b1789f84f820
  21:     0x563e6d0e2aa8 - helix_view::editor::Editor::launch_language_server::h60a8cc9881cf5ee6
  22:     0x563e6d0e511c - helix_view::editor::Editor::open::h1efe0893564dd5bd
  23:     0x563e6cdcb419 - helix_term::commands::lsp::jump_to_location::hc377bbf14fd09b18
  24:     0x563e6cdd1fe6 - helix_term::commands::lsp::goto_impl::heaf0097bdea53383
  25:     0x563e6d004fbb - core::ops::function::FnOnce::call_once{{vtable.shim}}::h579cfdb0393221e0
  26:     0x563e6ce7e831 - helix_term::job::Jobs::handle_callback::h4918681b88a5aaad
  27:     0x563e6d1b85ab - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::hbfb464632ff5164b
  28:     0x563e6d159723 - tokio::runtime::park::CachedParkThread::block_on::h8fb31feee02e3eeb
  29:     0x563e6d196747 - tokio::runtime::scheduler::multi_thread::MultiThread::block_on::hec5847f053998419
  30:     0x563e6d1bdad6 - hx::main::h23c91916c4a5df24
  31:     0x563e6d1a8933 - std::sys_common::backtrace::__rust_begin_short_backtrace::hc92a77c5c5557790
  32:     0x563e6d161ffd - std::rt::lang_start::{{closure}}::hc397c723102811d8
  33:     0x563e6d3017bb - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::had4f69b3aefb47a8
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/ops/function.rs:259:13
  34:     0x563e6d3017bb - std::panicking::try::do_call::hf2ad5355fcafe775
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:492:40
  35:     0x563e6d3017bb - std::panicking::try::h0a63ac363423e61e
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:456:19
  36:     0x563e6d3017bb - std::panic::catch_unwind::h18088edcecb8693a
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panic.rs:137:14
  37:     0x563e6d3017bb - std::rt::lang_start_internal::{{closure}}::ha7dad166dc711761
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/rt.rs:128:48
  38:     0x563e6d3017bb - std::panicking::try::do_call::hda0c61bf3a57d6e6
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:492:40
  39:     0x563e6d3017bb - std::panicking::try::hbc940e68560040a9
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:456:19
  40:     0x563e6d3017bb - std::panic::catch_unwind::haed0df2aeb3fa368
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panic.rs:137:14
  41:     0x563e6d3017bb - std::rt::lang_start_internal::h9c06694362b5b80c
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/rt.rs:128:20
  42:     0x563e6d1bdcb2 - main
  43:     0x7fc29b9825b0 - __libc_start_call_main
  44:     0x7fc29b982679 - __libc_start_main@GLIBC_2.2.5
  45:     0x563e6c9ea555 - _start
                               at /home/abuild/rpmbuild/BUILD/glibc-2.36/csu/../sysdeps/x86_64/start.S:115
  46:                0x0 - <unknown>
```

jknoetzke commented:
Thanks for the help on creating a build to test.
5748 is to allow LSP to work when there are multiple hierarchies of directories to a project ?
If so, I am testing using: https://github.com/awslabs/aws-lambda-rust-runtime/blob/main/examples/http-dynamodb/src/main.rs and LSP is not being engaged.

pascalkuthe commented:
> Thanks for the help on creating a build to test. 5748 is to allow LSP to work when there are multiple hierarchies of directories to a project ? If so, I am testing using: [awslabs/aws-lambda-rust-runtime@`main`/examples/http-dynamodb/src/main.rs](https://github.com/awslabs/aws-lambda-rust-runtime/blob/main/examples/http-dynamodb/src/main.rs?rgh-link-date=2023-02-07T16%3A44%3A44Z) and LSP is not being engaged.

You need to create a `.helix/config.toml` file at the top of the repo as shown in the PR description:
```toml
[editor]
workspace-lsp-roots = ["examples"]
```

pascalkuthe commented:
Thanks for reporting that @LeoniePhiline, I hate that tokio doesn't let me block on a `RwLock` in an async context. To fix that I would need to mark a ton of functions async. I just switched to a `parking_lot::RwLock` for now which always blocks but never panics like that. It should be fine as the lock is only held very briefly.

I also noticed that RA doesn't support multiple workspaces yet which is unfortunate and makes it behave slightly worse then it could. Specifically a new LSP is spawned for dependencies (seperate from the workspace). I implemented a fallback for the case that this LSP capability is not available so that in that case files outside of the workspace use the workspace LSP (instead of insisting on an lsp with `root_uri == None`) which should avoid spawning an additional LSP most of the time

pascalkuthe commented:
I submitted a patch to rust-analyzer (rust-lang/rust-analyzer#14098) to add support for `DidWorkspaceFoldersChanges` notifications. The patched RA version together with this PR work very well for rust.

pascalkuthe commented:
I just cameup with a solution to better support LSPs that don't support `DidWorkspaceChange` and multiple workspace folders (the servers that support all of these will work the best but I think I can comeuo with better behavior for other servers).

I don't have time right now but I will try to implement those improvements soon so that we can avoid spawning multiple LS whenever possible.

dead10ck commented on helix-lsp/src/client.rs:
```diff
@@ -25,6 +30,17 @@ use tokio::{
+        name: uri
+            .path_segments()
+            .and_then(|segments| segments.last())
+            .map(|basename| basename.to_string())
```
What do we do if there are conflicting workspace names?

pascalkuthe replied:
Workspace names are unimportant. RA for example.ignores them completely (most LS do). The language standard recommends using the base name (what we do here) as a default. LS only use these names in some messages: Like `indexing foo`

dead10ck commented on helix-lsp/src/client.rs:
```diff
@@ -36,11 +52,81 @@ pub struct Client {
+        if let Some(workspace_folders_caps) = self
+            .capabilities
+            .get()
+            .expect("Capabilities are known after initalization")
```
Should we just stop and print an error if we don't have the capabilities, rather than crashing?

pascalkuthe replied:
I simply inlined the `capabilities` function here which does the same thing. The capabilities are always filled on initialization, currently I don't handle uninitialized LSPs (see the todo a bit earlier). I habe some ideas how to handle that more gracefully (related to the changes I am talking about here: https://github.com/helix-editor/helix/pull/5748#issuecomment-1421637983, I don't habe time to work in that right now but once I do, I will hopefully also treat preinitialization servers more gracefully.

pascalkuthe replied:
unitalzied LSPs are now handeled properly

dead10ck commented on helix-term/src/config.rs:
```diff
@@ -33,6 +40,12 @@ pub enum ConfigLoadError {
 
+impl Default for ConfigLoadError {
+    fn default() -> Self {
+        ConfigLoadError::Error(IOError::new(std::io::ErrorKind::NotFound, "place holder"))
```
Did you mean this to be "place holder"?

pascalkuthe replied:
Yeah this is only used for testing and eben there only acts as a placeholder error for loading local config (all tests use global config). See `Config::load_test`

pascalkuthe commented:
> I just cameup with a solution to better support LSPs that don't support `DidWorkspaceChange` and multiple workspace folders (the servers that support all of these will work the best but I think I can comeuo with better behavior for other servers).
> 
> I don't have time right now but I will try to implement those improvements soon so that we can avoid spawning multiple LS whenever possible.

this change actually didn't workout as I had hoped. It works well enough for LSPs that don;'t support multiple workspaces by spawning new servers incase a root doesn't resolve to any existing server. However this can lead to (rare) cases where an addition server is needlesly spawned. The right fix here is to patch the servers to support this part of the LSP spec as I did for RA.

I also improved the way we deal with unitalized LSP clients (new roots for those LSP would be ignored until initalized) altough that also only works for servers that support multiple roots.

dead10ck reviewed (dismissed):
Ok this looks good to me!

LeoniePhiline commented:
I've just had a panic while using c1abca84:

```
thread 'main' panicked at 'Position 7396 is out of range for changeset len 0!', helix-core/src/transaction.rs:397:13
stack backtrace:
   0: rust_begin_unwind
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/panicking.rs:142:14
   2: helix_core::transaction::ChangeSet::map_pos
   3: helix_core::selection::Selection::map
   4: helix_view::view::View::apply
   5: helix_view::document::Document::append_changes_to_history
   6: core::ops::function::FnOnce::call_once{{vtable.shim}}
   7: helix_term::job::Jobs::handle_callback
   8: helix_term::application::Application::run::{{closure}}
   9: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
  10: tokio::runtime::park::CachedParkThread::block_on
  11: tokio::runtime::scheduler::multi_thread::MultiThread::block_on
  12: hx::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

I had prototyped in `src/main.rs` of the given project, then was ready to move everything over to `src/lib.rs` and put the former `fn main` into `mod tests`. At some point, `src/main.rs` was entirely empty and I saved both files - then hx crashed. I can't be sure if it had to do with the saving or some other async task.

pascalkuthe commented:
> I've just had a panic while using [c1abca8](https://github.com/helix-editor/helix/commit/c1abca847940aa2a385f0cc23a922d0527c8af08):
> 
> ```
> thread 'main' panicked at 'Position 7396 is out of range for changeset len 0!', helix-core/src/transaction.rs:397:13
> stack backtrace:
>    0: rust_begin_unwind
>              at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/panicking.rs:584:5
>    1: core::panicking::panic_fmt
>              at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/panicking.rs:142:14
>    2: helix_core::transaction::ChangeSet::map_pos
>    3: helix_core::selection::Selection::map
>    4: helix_view::view::View::apply
>    5: helix_view::document::Document::append_changes_to_history
>    6: core::ops::function::FnOnce::call_once{{vtable.shim}}
>    7: helix_term::job::Jobs::handle_callback
>    8: helix_term::application::Application::run::{{closure}}
>    9: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
>   10: tokio::runtime::park::CachedParkThread::block_on
>   11: tokio::runtime::scheduler::multi_thread::MultiThread::block_on
>   12: hx::main
> note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
> ```
> 
> I had prototyped in `src/main.rs` of the given project, then was ready to move everything over to `src/lib.rs` and put the former `fn main` into `mod tests`. At some point, `src/main.rs` was entirely empty and I saved both files - then hx crashed. I can't be sure if it had to do with the saving or some other async task.
> 
> The stack trace kinda looks like the panic might even be unrelated to the PR's changes?
> 
> _Edit:_ I also have very frequent RA LSP timeouts, e.g. when `:w`. (RA formatting times out, and the file is not being saved. I do `:lsp-restart pretty often`.
> 
> Might have to do with me using `askama_derive`?
> Helix log showing recurring `Async job failed: request timed out` and similar LSP timeouts

both of those issues seem fairly generic. The first crash commonly occurs when we forgot a history commit. It seems very unlikely to me that this PR would cause something like this. 

Same for the RA timeout, especially since you mention proc-macros this might be a RA issue rather than a helix issue. If you are able to reproduce this reliably then you could check if the same problem occurs with master.

LeoniePhiline commented:
> The first crash commonly occurs when we forgot a history commit.

**_If_** the async job timeouts have something to do with it, could it then be that the history save point was not committed in time -- waiting for an LSP async job to finish -- thus losing the race against some other component?

---

I switched back to master (as of c71b4c5579) and I am no longer getting the following two log messages, which constantly repeated (especially the first one - the other one should only occur on `:w`) while working with the [PR build](https://github.com/helix-editor/helix/commit/c1abca847940aa2a385f0cc23a922d0527c8af08):

```
helix_view::editor [ERROR] editor error: Async job failed: request timed out
helix_view::document [WARN] LSP formatting failed: request timed out
```
Maybe maybe it has something to do with the PR... I'll stay on the watch and keep `tail`ing the logs to get a better grasp on it.

In the coming days, I am going to keep switching back and forth `pull/5748/head` and `master`, to see if there is any regularity indicating causality, rather than coincidence.

---

Context: I am using RA rust-analyzer 0.3.1394-standalone (released 6 days ago).

pascalkuthe commented:
> > The first crash commonly occurs when we forgot a history commit.
> 
> **_If_** the async job timeouts have something to do with it, could it then be that the history save point was not committed in time -- waiting for an LSP async job to finish -- thus losing the race against some other component?
> 
> I switched back to master (as of [c71b4c5](https://github.com/helix-editor/helix/commit/c71b4c5579fdc753e62300f5ee6ca5e747a98e13)) and I am no longer getting the following two log messages, which constantly repeated (especially the first one - the other one should only occur on `:w`) while working with the [PR build](https://github.com/helix-editor/helix/commit/c1abca847940aa2a385f0cc23a922d0527c8af08):
> 
> ```
> helix_view::editor [ERROR] editor error: Async job failed: request timed out
> helix_view::document [WARN] LSP formatting failed: request timed out
> ```
> 
> Maybe maybe it has something to do with the PR... I'll stay on the watch and keep `tail`ing the logs to get a better grasp on it.
> 
> In the coming days, I am going to keep switching back and forth `pull/5748/head` and `master`, to see if there is any regularity indicating causality, rather than coincidence.
> 
> Context: I am using RA rust-analyzer 0.3.1394-standalone (released 6 days ago).

I am using RA build from source with my PR that adds better multi workspace support. That fix should be included in the release tomorrow. Perhaps something is not dealing well with the fact that there are multiple LSPs of the same kind. Or RA in particular doesn't deal with that well.

gibbz00 commented:
Also solved #5919 in #5636...

pascalkuthe commented:
@LeoniePhiline Did the recent RA update resolve your issues?

Also do these timeouts also appear for other LS or was it just RA specific?

LeoniePhiline commented:
I am thinking it might be a pure RA problem. I had been using `rust-analyzer 0.3.1394-standalone` with helix `master`, and yesterday I just couldn't save a file because RA would never come back and do its pre-save stuff (formatting etc.). I really wished we had #5964 already. :D

I have just now updated to `rust-analyzer 0.3.1402-standalone`. Let's see if that fixes anything.

Helix might want to deal better with misbehaving LSes, e.g. save files (and confirm to the user that the file is saved) even if the LS fails or times out.

I have not recently worked with other LSes. Previously, with `intelephense`, the problems I remember all had to do with the workspace and it not finding libraries. (The problem with `.git` folders stopping the workspace root search, so that the LS remains restricted to the sub-package, rather than the project root with all dependencies.)

pascalkuthe commented:
That's very odd I usually don't have issues with RA and I use it daily with many different projects. 
This might be a case of RA and cargo fighting jver the build directory lock. In that case it does sound like it's probably unrelated to this PR tough.

LeoniePhiline commented:
That might be (at least in the one project I worked on yesterday - where issues were more severe), although I have taken steps (insufficient steps?) to keep RA and cargo apart.
RA runs on the host (my laptop), while cargo watch runs in a container. The target dirs and cargo home dirs are separated, though.

In the container `CARGO_HOME=/.cargo`, and `cargo watch` is run as `CARGO_TARGET_DIR="/.target" cargo watch --why -B 1 -x run`. I would have expected that this resolves RA and cargo clashes.

Today I'm going back to working on a pure CLI project with no container or otherwise `cargo watch` involved. I will observe if there are issues using helix `master`. If there are none, then I will switch to the PR branch to see if any issues are added.

My apologies if I have not been as straightforward as desired with pinpointing and narrowing down the issues and their causes. There are many moving parts, and I need to take them all into account and observe carefully (while doing the actual work). As I learn more about how editors and LSes interact, as well as how different rust tools interact, I will gain a clearer picture and be able to debug these things with a sharper knife.

the-mikedavis commented on book/src/languages.md:
```diff
@@ -64,6 +64,7 @@ These configuration keys are available:
 | `grammar`             | The tree-sitter grammar to use (defaults to the value of `name`) |
 | `formatter`           | The formatter for the language, it will take precedence over the lsp when defined. The formatter must be able to take the original file as input from stdin and write the formatted file to stdout |
 | `text-width`          |  Maximum line length. Used for the `:reflow` command and soft-wrapping if `soft-wrap.wrap_at_text_width` is set, defaults to `editor.text-width`   |
+| `workspace-lsp-roots`     | Directories relative to the workspace root that are treated as LSP roots. Should only be set in `.helix/config.toml`. Overwrites the setting of the same name in `config.toml` if set. | `` |
```
```suggestion
| `workspace-lsp-roots`     | Directories relative to the workspace root that are treated as LSP roots. Should only be set in `.helix/config.toml`. Overwrites the setting of the same name in `config.toml` if set. |
```

the-mikedavis commented on helix-lsp/src/client.rs:
```diff
@@ -36,11 +51,126 @@ pub struct Client {
+                    .any(|workspace| &workspace.uri == root_uri)
+            })
+        {
+            // workspace URI is already register we can use this client
```
```suggestion
            // workspace URI is already registered so we can use this client
```

the-mikedavis commented on helix-lsp/src/client.rs:
```diff
@@ -36,11 +51,126 @@ pub struct Client {
+            .push(workspace_for_uri(root_uri.clone()));
+        if &Some(OneOf::Left(false)) == change_notifications {
+            // server specifically opted out of DidWorkspaceChange notifications
+            // let's assume the server will request the workspace folders himself
```
```suggestion
            // let's assume the server will request the workspace folders itself
```

the-mikedavis commented on helix-lsp/src/lib.rs:
```diff
@@ -540,41 +548,45 @@ impl Registry {
     pub fn stop(&mut self, language_config: &LanguageConfiguration) {
         let scope = language_config.scope.clone();
 
-        if let Some((_, client)) = self.inner.remove(&scope) {
-            tokio::spawn(async move {
-                let _ = client.force_shutdown().await;
-            });
+        if let Some(clients) = self.inner.remove(&scope) {
+            for (_, client) in clients {
+                tokio::spawn(async move {
```
I was curious if it's better here to have the `for` inside or outside the `spawn`. Any risk of bad things happening from `tokio::spawn`ing too many futures?

pascalkuthe replied:
Compared to spawning threads spawning tokio tasks is very cheap. In fact I have read that tokio is able to scale up to hundreds of thousands of tasks in production servers. From the docs:

> Tasks are light weight. Because tasks are scheduled by the Tokio runtime rather than the operating system, creating new tasks or switching between tasks does not require a context switch and has fairly low overhead. Creating, running, and destroying large numbers of tasks is quite cheap, especially compared to OS threads.

The idea is basically that an async tokio task spends most of its time `await`ing something so having a lot of them has little overhead because they don't actually do anything most of the time. The nice thing about async io is that the task gets automatically woken up (and queued to be run) once the io is finished so until then `tokio` only needs to store that task [^1]. In the worst case (single threaded executer) tokio essentially runs these tasks consequitvely in a loop so that would be about equivalent to moving the for loop insidie the task (because we await anyway).

This case is a classical example of: spawn a task that does very little computation and instead just sends a message and waits for a response so it just felt right to do it this way. In the (very weird/unlikely) case that we have a large number of servers that all timeout while sending responses this approach is somewhat more optimal because the servers are shutdown in parallel instead of consecutively.
 
That said moving the for loop inside the task works too and there is likely little difference either way. I am certain that we will never reach the point where the overhead of spawning more tokio tasks becomes relevant. An editor just doesn't spawn that many tasks compared to a server :D

I did notice that I was inconsistent here and had the for loop inside the spawn in `restart` I moved that outside too so it's atleast consistent and looks less jarring.

[^1]: That means we should make sure that all "heavy" computations (for example diffs, updating the document running TS) computation are not done in tokio tasks but inside a blocking context instead: see [`block_in_place`](https://docs.rs/tokio/latest/tokio/task/fn.block_in_place.html) and [`spawn_blocking`](https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html)

the-mikedavis reviewed (commented):
I've been running this for a while and it works great! I have some minor comments

pascalkuthe commented on helix-lsp/src/client.rs:
```diff
@@ -36,11 +51,126 @@ pub struct Client {
+        // TODO refactor to use let .. else when msrv is raised to 1.65
+        let capabilities = if let Some(capabilities) = self.capabilities.get() {
+            capabilities
+        } else {
```
thanks forgot about that :)

the-mikedavis commented on helix-term/src/commands/typed.rs:
```diff
@@ -2601,6 +2618,13 @@ pub const TYPABLE_COMMAND_LIST: &[TypableCommand] = &[
+            aliases: &[],
+            doc: "Open the workspace config.toml file.",
+            fun: open_workspace_config,
+            completer: None,
```
With #5966 merged this should become

```suggestion
            signature: CommandSignature::none(),
```

to prevent a breakage on master

pascalkuthe replied:
thanks! I only superficially check that this has no conflicts but this only shows up as a compiler error after merge. I will rebase on master to make sure there is nothing else wrong. I should setup a bot that automatically rebases my PRs regularly and notifies me if there are conflicts or CI fails :D

krobelus commented:
> If there is not LSP yet: Spawn an LSP with workspace_dirs set to an empty list, root_uri set to None and root_dir set to CWD (see https://github.com/helix-editor/helix/issues/4436#issuecomment-1288802003, root_dir must be set for backwards compatibility)

@pascalkuthe is this still the behavior?
I'm trying to fix the same marksman issue in Kakoune.
It looks like Helix always adds the file's root as single workspace folder when starting a new server:
In `helix-lsp/src/client.rs`
```
        let workspace_folders = root_uri
            .clone()
            .map(|root| vec![workspace_for_uri(root)])
            .unwrap_or_default();
```
